pkg/controller/networkpolicyrecommendation/controller.go

```diff
@@ -206,6 +215,95 @@ func (c *NPRecommendationController) Run(stopCh <-chan struct{}) {
 	<-stopCh
 }
 
+// handleStaleResources handles the stale Spark Applications and database entries.
+// It will delete the dangling resources without a matching NetworkPolicyRecommendation
+// and add the running NetworkPolicyRecommendation back to the periodical watch list.
+func (c *NPRecommendationController) handleStaleResources() error {
+	if c.clickhouseConnect == nil {
+		var err error
+		c.clickhouseConnect, err = clickhouse.SetupConnection(&c.kubeClient)
+		if err != nil {
+			return fmt.Errorf("failed to connect ClickHouse: %v", err)
+		}
+	}
+	idList, err := getPolicyRecommendationIds(c.clickhouseConnect)
+	if err != nil {
+		return fmt.Errorf("failed to get recommendation ids from ClickHouse: %v", err)
+	}
+	var errorList []error
+	// Remove stale database entries
+	for _, id := range idList {
+		_, err := c.GetNetworkPolicyRecommendation(env.GetTheiaNamespace(), "pr-"+id)
+		if err != nil {
+			if apimachineryerrors.IsNotFound(err) {
+				err = deletePolicyRecommendationResult(c.clickhouseConnect, id)
+				if err != nil {
+					errorList = append(errorList, err)
+				}
+			} else {
+				errorList = append(errorList, err)
+			}
+		}
+	}
+
+	saList, err := ListSparkApplication(c.kubeClient, sparkAppLabel)
+	if err != nil {
+		return fmt.Errorf("failed to list Spark Application: %v", err)
+	}
+	// Remove stale Spark Applications
+	for _, sa := range saList.Items {
+		_, err := c.GetNetworkPolicyRecommendation(sa.Namespace, sa.Name)
+		if err != nil {
+			if apimachineryerrors.IsNotFound(err) {
+				DeleteSparkApplication(c.kubeClient, sa.Name, sa.Namespace)
+			} else {
+				errorList = append(errorList, err)
+			}
+		}
+	}
+	if len(errorList) > 0 {
+		return fmt.Errorf("failed to remove stale Spark Applications and database entries: %v", errorList)
+	}
+
+	// Add scheduled/running NPR back to resycn list
+	nprList, err := c.ListNetworkPolicyRecommendation(env.GetTheiaNamespace())
```

Comment:
Could there be a race case when a job is deleted between `ListNetworkPolicyRecommendation` and  `c.addPeriodicSync` is called? Shouldn't be of much issue though as even that's the case, during the next resync the cleanup can be done w/o issue.

Reply:
I think it is fine as the PeriodicSync just add the NPR key to the main workqueue, `syncNPRecommendation` will return nil if it does not find the NPR. As the theia manager is on, the deletion process will take care of the SparkApplication, db entires and also PeriodicSyncSet.